### DISCUSSION TITLE
Basic doors

### DISCUSSION THREAD

**Discussion by degradka** [jitspoe/godot_bsp_importer#19] [Q&A] [answered]
I'm kinda stupid, so I need help with making doors. I made a simple solid class func_door in the fgd, made a brush func_door and remapped it to a characterbody3d. But when I try to apply some rotation magic, the pivot of the door is at world's center, so the door rotates weirdly.
P.S. Would be great if somebody made a bunch of example entities for the addon.

**Comment by jitspoe** [comment, marked as answer]
Yeah, that's how the Quake BSP works as far as world brushes go, BUT you can do some trickery to have rotating doors in Quake: https://www.slipseer.com/index.php?threads/top-10-mapper-hints.186/#post-1227

tl;dr version is you have a "rotate_object" (which is the brush-based entity) which targets an "info_rotate" (a point-based entity).  I *think* the compiler then sets the offset of the entity to match the "info_rotate", but worst case you can do some math to pivot around the point entity location.

Quake2 made that simpler by simply having a brush with the "origin" texture that you'd place where the door hinge should be.  I don't have the Quake2 format supported yet, though.  @Yagich was interested in perhaps finishing support for that, though.

**Reply by degradka** [reply]
Thanks for the answer! Yeah, Quake2 format sounds like a no pain solution for this.

**Comment by hemebond** [comment]
You can use the origin brush trick with q1bsp too.

<img width="811" height="636" alt="image" src="https://github.com/user-attachments/assets/52fc05dd-ecf8-421a-af90-4569cba7a898" />

<img width="811" height="636" alt="image" src="https://github.com/user-attachments/assets/bb5fc3db-3e76-4db0-a279-761a05749604" />

**Reply by degradka** [reply]
Yeah, damn, I totally forgot about it

**Reply by jitspoe** [reply]
Oh, interesting.  I had no idea that was possible, but I guess it's more of a compiler thing than a format thing, so it makes sense that modern compilers could do that.

**Reply by hemebond** [reply]
It would still be really, _really_ good if the BSP importer could set the origin of brush entities, in the absence of an origin brush, to the center of the entity collision shapes. And perhaps also make a good AABB available somehow. There are rotations on brush entities that make it difficult for me to calculate the AABB in the post script.

**Reply by jitspoe** [reply]
I've added helper functions, so you can call var mins_maxs := BSPReader.get_brush_entity_mins_maxs(your_entity_node).  Also one for getting the center, ex:

```gdscript
func post_import(root_node : Node):
	var sound_location := BSPReader.get_brush_entity_center(self)
	var sound_loop_player : AudioStreamPlayer3D = preload("res://sound/machinery/sound_gate_engine_loop.tscn").instantiate()
	add_child(sound_loop_player)
```
